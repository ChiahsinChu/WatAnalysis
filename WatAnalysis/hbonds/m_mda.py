# SPDX-License-Identifier: LGPL-3.0-or-later
import numpy as np
from MDAnalysis.analysis.hydrogenbonds.hbond_analysis import HydrogenBondAnalysis
from MDAnalysis.lib.distances import calc_angles, capped_distance


class PartialHBAnalysis(HydrogenBondAnalysis):
    def __init__(
        self,
        universe,
        hb_region=None,
        surf_ids=None,
        donors_sel=None,
        hydrogens_sel=None,
        acceptors_sel=None,
        between=None,
        d_h_cutoff=1.2,
        d_a_cutoff=3.0,
        angle_cutoff_type="d_h_a",
        angle_cutoff=None,
        update_selections=True,
        update_masks=False,
    ):
        """
        Parameters
        ----------
        hb_region : (2, )-shape List (None)
            region (z positions w.r.t surface) in which the HB is analyzed
            if None, all HB is considered
        surf_ids : (2, n_surf)-shape List (None)
            atomic indices of surface atoms
            expected to be not None if region is not None
        angle_cutoff_type : d_h_a or h_d_a
            type for setting angle cutoff
            d_h_a: donor-hydrogen-acceptor
            h_d_a: hydrogen-donor-acceptor
        update_masks : bool (False)
            if calculate the masks for donors & acceptors individually
            False if donors & acceptors are exactly the same
            and then share their mask
        Other parameters are identical with their counterparts in the parent class.
        """

        # set new cutoff attributes
        self.angle_cutoff_type = angle_cutoff_type
        if self.angle_cutoff_type == "d_h_a":
            if angle_cutoff is None:
                self.angle_cutoff = 150
            else:
                self.angle_cutoff = angle_cutoff
        elif self.angle_cutoff_type == "h_d_a":
            if angle_cutoff is None:
                self.angle_cutoff = 30
            else:
                self.angle_cutoff = angle_cutoff
        else:
            raise AttributeError("Unknown angle cutoff type!")
        self.update_masks = update_masks

        super(PartialHBAnalysis, self).__init__(
            universe,
            donors_sel,
            hydrogens_sel,
            acceptors_sel,
            between,
            d_h_cutoff,
            d_a_cutoff,
            self.angle_cutoff,
            update_selections,
        )

        self.surf_ids = surf_ids
        if surf_ids is not None:
            # TODO: universe should be wrapped before passed
            self.surf_ids = np.array(surf_ids, dtype=np.int32)

        self.hb_region = hb_region
        if hb_region is not None:
            self.hb_region = np.array(self.hb_region)
            if self.hb_region.shape != (2,):
                raise AttributeError("hb_region is expected to be a (2, )-shape array")
            if self.surf_ids is None:
                raise AttributeError(
                    "surf_ids is expected to be not None when hb_region is not None"
                )

        # trajectory value initial
        self.ag = universe.atoms
        self.ag_natoms = len(self.ag)

        # parallel value initial
        self.para = None
        self._para_region = None

    def _prepare(self):
        """
        Add z coords of donors/hydrogens/acceptors w.r.t. the
        nearest surface into results

        self.results.hbonds:
        frame_idx, donor_idx, hydrogen_idx, acceptor_idx, HB distance,
        HB angle, donor_zcoord, hydrogen_zcoord, acceptor_zcoord
        """
        super(PartialHBAnalysis, self)._prepare()
        self.results.hbonds = [[], [], [], [], [], [], [], [], []]
        self.zs_surf = np.zeros((self.n_frames, 2), dtype=np.float32)

    def _single_frame(self):
        box = self._ts.dimensions

        # Update donor-hydrogen pairs if necessary
        if self.update_selections:
            self._donors, self._hydrogens = self._get_dh_pairs()
        donors = self._donors
        hydrogens = self._hydrogens
        acceptors = self._acceptors

        # get selected AtomGroup donors, hydrogens, acceptors
        if self.hb_region is not None:
            ts_surf_zlo = self._ts.positions.T[2][self.surf_ids[0]]
            ts_surf_zhi = self._ts.positions.T[2][self.surf_ids[1]]
            z_surf = [np.mean(ts_surf_zlo), np.mean(ts_surf_zhi)]
            np.copyto(self.zs_surf[self._frame_index], z_surf)
            # print(z_surf)

            # Mask to select atoms at specified regions
            mask = self._get_mask(z_surf, self._donors.positions)
            donors = self._donors[mask]
            hydrogens = self._hydrogens[mask]
            left_donors = self._donors[mask == False]
            left_hydrogens = self._hydrogens[mask == False]
            if self.update_masks == False:
                mask = mask[::2]
            else:
                mask = self._get_mask(z_surf, self._acceptors.positions)
            acceptors = self._acceptors[mask]

        # find D (selected) and A (all) within cutoff distance of one another
        # min_cutoff = 1.0 as an atom cannot form a hydrogen bond with itself
        d_a_indices, d_a_distances = capped_distance(
            donors.positions,
            self._acceptors.positions,
            max_cutoff=self.d_a_cutoff,
            min_cutoff=1.0,
            box=box,
            return_distances=True,
        )

        # Remove D-A pairs more than d_a_cutoff away from one another
        tmp_donors = donors[d_a_indices.T[0]]
        tmp_hydrogens = hydrogens[d_a_indices.T[0]]
        tmp_acceptors = self._acceptors[d_a_indices.T[1]]

        # Remove donor-acceptor pairs between pairs of AtomGroups we are not
        # interested in
        if self.between_ags is not None:
            tmp_donors, tmp_hydrogens, tmp_acceptors = self._filter_atoms(
                tmp_donors, tmp_hydrogens, tmp_acceptors
            )

        # Find angles and compared with the angle cutoff
        if self.angle_cutoff_type == "d_h_a":
            cutoff_angles = np.rad2deg(
                calc_angles(
                    tmp_donors.positions,
                    tmp_hydrogens.positions,
                    tmp_acceptors.positions,
                    box=box,
                )
            )
            hbond_indices = np.where(cutoff_angles >= self.angle_cutoff)[0]
        elif self.angle_cutoff_type == "h_d_a":
            cutoff_angles = np.rad2deg(
                calc_angles(
                    tmp_hydrogens.positions,
                    tmp_donors.positions,
                    tmp_acceptors.positions,
                    box=box,
                )
            )
            hbond_indices = np.where(cutoff_angles <= self.angle_cutoff)[0]

        # Retrieve atoms, distances and angles of hydrogen bonds
        hbond_donors = tmp_donors[hbond_indices]
        hbond_hydrogens = tmp_hydrogens[hbond_indices]
        hbond_acceptors = tmp_acceptors[hbond_indices]
        hbond_distances = d_a_distances[hbond_indices]
        hbond_angles = cutoff_angles[hbond_indices]
        if self.hb_region is not None:
            donor_zcoords = self._get_rel_pos(hbond_donors.positions[:, 2], z_surf)
            hydrogen_zcoords = self._get_rel_pos(
                hbond_hydrogens.positions[:, 2], z_surf
            )
            acceptor_zcoords = self._get_rel_pos(
                hbond_acceptors.positions[:, 2], z_surf
            )

        # Store data on hydrogen bonds found at this frame
        self.results.hbonds[0].extend(np.full_like(hbond_donors, self._ts.frame))
        self.results.hbonds[1].extend(hbond_donors.indices)
        self.results.hbonds[2].extend(hbond_hydrogens.indices)
        self.results.hbonds[3].extend(hbond_acceptors.indices)
        self.results.hbonds[4].extend(hbond_distances)
        self.results.hbonds[5].extend(hbond_angles)

        if self.hb_region is None:
            placeholder = np.zeros_like(hbond_angles).tolist()
            self.results.hbonds[6].extend(placeholder)
            self.results.hbonds[7].extend(placeholder)
            self.results.hbonds[8].extend(placeholder)
        else:
            self.results.hbonds[6].extend(donor_zcoords)
            self.results.hbonds[7].extend(hydrogen_zcoords)
            self.results.hbonds[8].extend(acceptor_zcoords)
            # consider D(not selected)-A(selected) pair
            d_a_indices, d_a_distances = capped_distance(
                left_donors.positions,
                acceptors.positions,
                max_cutoff=self.d_a_cutoff,
                min_cutoff=1.0,
                box=box,
                return_distances=True,
            )

            # Remove D-A pairs more than d_a_cutoff away from one another
            tmp_donors = left_donors[d_a_indices.T[0]]
            tmp_hydrogens = left_hydrogens[d_a_indices.T[0]]
            tmp_acceptors = acceptors[d_a_indices.T[1]]

            # Remove donor-acceptor pairs between pairs of AtomGroups we are not
            # interested in
            if self.between_ags is not None:
                tmp_donors, tmp_hydrogens, tmp_acceptors = self._filter_atoms(
                    tmp_donors, tmp_hydrogens, tmp_acceptors
                )

            # Find angles and compared with the angle cutoff
            if self.angle_cutoff_type == "d_h_a":
                cutoff_angles = np.rad2deg(
                    calc_angles(
                        tmp_donors.positions,
                        tmp_hydrogens.positions,
                        tmp_acceptors.positions,
                        box=box,
                    )
                )
                hbond_indices = np.where(cutoff_angles > self.angle_cutoff)[0]
            elif self.angle_cutoff_type == "h_d_a":
                cutoff_angles = np.rad2deg(
                    calc_angles(
                        tmp_hydrogens.positions,
                        tmp_donors.positions,
                        tmp_acceptors.positions,
                        box=box,
                    )
                )
                hbond_indices = np.where(cutoff_angles < self.angle_cutoff)[0]

            # Retrieve atoms, distances and angles of hydrogen bonds
            hbond_donors = tmp_donors[hbond_indices]
            hbond_hydrogens = tmp_hydrogens[hbond_indices]
            hbond_acceptors = tmp_acceptors[hbond_indices]
            hbond_distances = d_a_distances[hbond_indices]
            hbond_angles = cutoff_angles[hbond_indices]
            donor_zcoords = self._get_rel_pos(hbond_donors.positions[:, 2], z_surf)
            hydrogen_zcoords = self._get_rel_pos(
                hbond_hydrogens.positions[:, 2], z_surf
            )
            acceptor_zcoords = self._get_rel_pos(
                hbond_acceptors.positions[:, 2], z_surf
            )

            # Store data on hydrogen bonds found at this frame
            self.results.hbonds[0].extend(np.full_like(hbond_donors, self._ts.frame))
            self.results.hbonds[1].extend(hbond_donors.indices)
            self.results.hbonds[2].extend(hbond_hydrogens.indices)
            self.results.hbonds[3].extend(hbond_acceptors.indices)
            self.results.hbonds[4].extend(hbond_distances)
            self.results.hbonds[5].extend(hbond_angles)
            self.results.hbonds[6].extend(donor_zcoords)
            self.results.hbonds[7].extend(hydrogen_zcoords)
            self.results.hbonds[8].extend(acceptor_zcoords)

    def _get_mask(self, z_surf, positions):
        """
        TBC
        """
        z_coords = positions[:, 2]
        abs_region = self._get_abs_region(self.hb_region, z_surf)
        # print(abs_region)
        mask_0 = z_coords >= abs_region[0][0]
        mask_1 = z_coords < abs_region[0][1]
        mask_2 = z_coords >= abs_region[1][0]
        mask_3 = z_coords < abs_region[1][1]
        mask = (mask_0 & mask_1) | (mask_2 & mask_3)
        return mask

    def _get_abs_region(self, hb_region, z_surf):
        """
        TBC
        """
        region_0 = z_surf[0] + hb_region
        region_1 = z_surf[1] - hb_region
        tmp = [region_1[1], region_1[0]]
        region_1 = np.array(tmp)
        abs_region = [region_0, region_1]
        return np.array(abs_region)

    def _get_rel_pos(self, positions, z_surf):
        """
        TBC
        """
        # z positions w.r.t. lower surface
        rel_pos_0 = positions - z_surf[0]
        # z positions w.r.t. upper surface
        rel_pos_1 = z_surf[1] - positions
        # take the shorter distance
        rel_pos = np.concatenate(([rel_pos_0], [rel_pos_1]), axis=0)
        # pos for lower, neg for upper
        mask = -2 * np.argmin(rel_pos, axis=0) + 1
        rel_pos = np.min(rel_pos, axis=0) * mask
        return rel_pos

    def _parallel_init(self, *args, **kwargs):
        start = self._para_region.start
        stop = self._para_region.stop
        step = self._para_region.step
        self._setup_frames(self._trajectory, start, stop, step)
        self._prepare()

    def run(self, start=None, stop=None, step=None, verbose=None):
        # self._trajectory._reopen()
        if verbose == True:
            print(" ", end="")
        super().run(
            start, stop, step, verbose
        )  ### will be problem for conclude operation

        if self.para:
            block_result = self._para_block_result()
            if block_result == None:
                raise ValueError(
                    "in parallel, block result has not been defined or no data output!"
                )
            # logger.info("block_anal finished.")
            return block_result

    def _para_block_result(
        self,
    ):
        return [self.results.hbonds]

    def _parallel_conclude(self, rawdata):
        # set attributes for further analysis
        method_attr = rawdata[-1]
        del rawdata[-1]
        self.start = method_attr[0]
        self.stop = method_attr[1]
        self.step = method_attr[2]
        self.frames = np.arange(self.start, self.stop, self.step)

        total_array = np.empty(shape=(0, rawdata[0][0].shape[1]))
        for single_data in rawdata:
            total_array = np.concatenate([total_array, single_data[0]], axis=0)

        self.results["hbonds"] = total_array

        return "FINISH PARA CONCLUDE"
