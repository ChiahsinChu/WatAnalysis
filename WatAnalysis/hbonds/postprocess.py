"""
Functions for postprocessing of PartialHBAnalysis results
"""
import numpy as np
import os


def count_by_time(hbonds_result, start, stop, step=1):
    """
    Adapted from MDA
    """
    indices, tmp_counts = np.unique(hbonds_result[:, 0],
                                    axis=0,
                                    return_counts=True)
    indices -= start
    indices /= step
    counts = np.zeros_like(np.arange(start, stop, step))
    counts[indices.astype(np.intp)] = tmp_counts
    return np.arange(start, stop, step), counts


def lifetime(hbonds_result,
             start,
             stop,
             step,
             dt,
             tau_max=20,
             window_step=1,
             intermittency=0):
    """
    Adapted from MDA
    """
    from MDAnalysis.lib.correlations import autocorrelation, correct_intermittency

    frames = np.arange(start, stop, step)
    found_hydrogen_bonds = [set() for _ in frames]
    for frame_index, frame in enumerate(frames):
        for hbond in hbonds_result[hbonds_result[:, 0] == frame]:
            found_hydrogen_bonds[frame_index].add(frozenset(hbond[2:4]))

    intermittent_hbonds = correct_intermittency(found_hydrogen_bonds,
                                                intermittency=intermittency)
    tau_timeseries, timeseries, timeseries_data = autocorrelation(
        intermittent_hbonds, tau_max, window_step=window_step)
    output = np.vstack([tau_timeseries, timeseries])
    output[0] = output[0] * dt
    return output


def fit_biexponential(tau_timeseries, ac_timeseries):
    """Fit a biexponential function to a hydrogen bond time autocorrelation function

    Return the two time constants

    Adapted from MDA
    """
    from scipy.optimize import curve_fit

    def model(t, A, tau1, B, tau2):
        """Fit data to a biexponential function.
        """
        return A * np.exp(-t / tau1) + B * np.exp(-t / tau2)

    params, params_covariance = curve_fit(model, tau_timeseries, ac_timeseries,
                                          [1, 0.5, 1, 2])

    fit_t = np.linspace(tau_timeseries[0], tau_timeseries[-1], 1000)
    fit_ac = model(fit_t, *params)

    return params, fit_t, fit_ac


def get_graphs(hbonds_result, start, stop, step, output_dir):
    output_dir = os.path.realpath(output_dir)
    ts, counts = count_by_time(hbonds_result, start)
    tmp_count = 0
    for t, count in zip(ts, counts):
        tmp_result = hbonds_result[tmp_count:tmp_count+count]
        output = get_graph(tmp_result)
        np.save(output, output_dir + "/" + str(t) + ".txt")
        tmp_count = tmp_count + count

        
def get_graph(tmp_result):
    output = 0
    return output