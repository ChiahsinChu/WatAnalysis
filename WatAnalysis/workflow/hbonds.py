# SPDX-License-Identifier: LGPL-3.0-or-later
from typing import Dict

import freud
import numpy as np
from ase import geometry
from MDAnalysis.lib.distances import (
    calc_angles,
    capped_distance,
    distance_array,
    minimize_vectors,
)

from WatAnalysis import utils
from WatAnalysis.workflow.base import (
    DataRequirement,
    PlanarInterfaceAnalysisBase,
    SingleAnalysis,
)


class HydrogenBondAnalysis(SingleAnalysis):
    def __init__(
        self,
        oxygen_sel: str = "name O",
        hydrogens_sel: str = "name H",
        label: str = "oxygen",
        OH_cutoff: float = 1.3,
        HB_cutoff: Dict[str, float] = None,
        max_HB: int = 4,
    ) -> None:
        super().__init__()

        self.oxygen_sel = oxygen_sel
        self.hydrogens_sel = hydrogens_sel

        self.label = label
        self.OH_cutoff = OH_cutoff
        # self.HB_cutoff = HB_cutoff
        self.DA_cutoff = HB_cutoff.get("DA", 3.5)
        self.HDA_cutoff = HB_cutoff.get("HDA", None)
        self.DHA_cutoff = HB_cutoff.get("DHA", None)
        assert (
            self.HDA_cutoff is not None or self.DHA_cutoff is not None
        ), "Either HDA_cutoff or DHA_cutoff must be provided."
        self.max_HB = max_HB

        self.data_requirements = {
            f"donor_z_{self.label}": DataRequirement(
                f"donor_z_{self.label}",
                atomic=True,
                dim=max_HB,
                selection=self.oxygen_sel,
                value=np.NaN,
            ),
            f"acceptor_z_{self.label}": DataRequirement(
                f"acceptor_z_{self.label}",
                atomic=True,
                dim=max_HB,
                selection=self.oxygen_sel,
                value=np.NaN,
            ),
            f"donor_cn_{self.label}": DataRequirement(
                f"donor_cn_{self.label}",
                atomic=True,
                dim=max_HB,
                selection=self.oxygen_sel,
                value=0,
                dtype=int,
            ),
            f"acceptor_cn_{self.label}": DataRequirement(
                f"acceptor_cn_{self.label}",
                atomic=True,
                dim=max_HB,
                selection=self.oxygen_sel,
                value=0,
                dtype=int,
            ),
            f"da_distance_{self.label}": DataRequirement(
                f"da_distance_{self.label}",
                atomic=True,
                dim=max_HB,
                selection=self.oxygen_sel,
                value=np.NaN,
            ),
            f"hbond_angle_{self.label}": DataRequirement(
                f"hbond_angle_{self.label}",
                atomic=True,
                dim=max_HB,
                selection=self.oxygen_sel,
                value=np.NaN,
            ),
        }

        self.ag_oxygen = None
        self.ag_hydrogen = None

    def _prepare(self, analyser: PlanarInterfaceAnalysisBase):
        self.ag_oxygen = analyser.universe.select_atoms(self.oxygen_sel)
        self.ag_hydrogen = analyser.universe.select_atoms(self.hydrogens_sel)

    def _single_frame(self, analyser: PlanarInterfaceAnalysisBase):
        ts_box = analyser._ts.dimensions

        # return [O_id, H_id] for each O-H pair
        ts_OH_pairs = capped_distance(
            self.ag_oxygen.positions,
            self.ag_hydrogen.positions,
            max_cutoff=self.OH_cutoff,
            box=ts_box,
            return_distances=False,
        )
        # find A-D pairs based on O-O distance
        _donors = self.ag_oxygen[ts_OH_pairs[:, 0]]
        _hydrogens = self.ag_hydrogen[ts_OH_pairs[:, 1]]
        ts_OO_pairs, DA_distances = capped_distance(
            _donors.positions,
            self.ag_oxygen.positions,
            max_cutoff=self.DA_cutoff,
            min_cutoff=1.0,
            box=ts_box,
            return_distances=True,
        )
        tmp_donors = _donors[ts_OO_pairs[:, 0]]
        tmp_hydrogens = _hydrogens[ts_OO_pairs[:, 0]]
        tmp_acceptors = self.ag_oxygen[ts_OO_pairs[:, 1]]
        # check the angle threshold is H-O...O or O-H...O
        # Find angles and compared with the angle cutoff
        if self.HDA_cutoff is not None:
            cutoff_angles = np.rad2deg(
                calc_angles(
                    tmp_hydrogens.positions,
                    tmp_donors.positions,
                    tmp_acceptors.positions,
                    box=ts_box,
                )
            )
            hbond_indices = np.where(cutoff_angles <= self.HDA_cutoff)[0]
        elif self.DHA_cutoff is not None:
            cutoff_angles = np.rad2deg(
                calc_angles(
                    tmp_donors.positions,
                    tmp_hydrogens.positions,
                    tmp_acceptors.positions,
                    box=ts_box,
                )
            )
            hbond_indices = np.where(cutoff_angles >= self.DHA_cutoff)[0]
        else:
            raise ValueError("Either HDA_cutoff or DHA_cutoff must be provided.")

        # Retrieve atoms, distances and angles of hydrogen bonds
        hbond_donors = tmp_donors[hbond_indices]
        hbond_acceptors = tmp_acceptors[hbond_indices]

        np.copyto(
            getattr(analyser, f"donor_z_{self.label}")[analyser._frame_index],
            np.pad(
                hbond_donors.positions[:, analyser.axis],
                (0, self.max_HB * self.ag_oxygen.n_atoms - hbond_donors.n_atoms),
                mode="constant",
                constant_values=np.nan,
            ).reshape(self.ag_oxygen.n_atoms, self.max_HB),
        )
        np.copyto(
            getattr(analyser, f"acceptor_z_{self.label}")[analyser._frame_index],
            np.pad(
                hbond_acceptors.positions[:, analyser.axis],
                (0, self.max_HB * self.ag_oxygen.n_atoms - hbond_acceptors.n_atoms),
                mode="constant",
                constant_values=np.nan,
            ).reshape(self.ag_oxygen.n_atoms, self.max_HB),
        )
        np.copyto(
            getattr(analyser, f"da_distance_{self.label}")[analyser._frame_index],
            np.pad(
                DA_distances[hbond_indices],
                (0, self.max_HB * self.ag_oxygen.n_atoms - hbond_donors.n_atoms),
                mode="constant",
                constant_values=np.nan,
            ).reshape(self.ag_oxygen.n_atoms, self.max_HB),
        )
        np.copyto(
            getattr(analyser, f"hbond_angle_{self.label}")[analyser._frame_index],
            np.pad(
                cutoff_angles[hbond_indices],
                (0, self.max_HB * self.ag_oxygen.n_atoms - hbond_donors.n_atoms),
                mode="constant",
                constant_values=np.nan,
            ).reshape(self.ag_oxygen.n_atoms, self.max_HB),
        )
        # calculate CN for each oxygen atom

        ts_cn = np.zeros(self.ag_oxygen.n_atoms, dtype=int)
        oxygen_ids, _cn = np.unique(ts_OH_pairs[:, 0], return_counts=True)
        ts_cn[oxygen_ids] = _cn
        full_cn = np.zeros(analyser.universe.atoms.n_atoms, dtype=int)
        full_cn[self.ag_oxygen.indices] = ts_cn
        # print(full_cn)
        np.copyto(
            getattr(analyser, f"donor_cn_{self.label}")[analyser._frame_index],
            np.pad(
                full_cn[hbond_donors.indices],
                (0, self.max_HB * self.ag_oxygen.n_atoms - hbond_donors.n_atoms),
                mode="constant",
                constant_values=0,
            ).reshape(self.ag_oxygen.n_atoms, self.max_HB),
        )
        np.copyto(
            getattr(analyser, f"acceptor_cn_{self.label}")[analyser._frame_index],
            np.pad(
                full_cn[hbond_acceptors.indices],
                (0, self.max_HB * self.ag_oxygen.n_atoms - hbond_acceptors.n_atoms),
                mode="constant",
                constant_values=0,
            ).reshape(self.ag_oxygen.n_atoms, self.max_HB),
        )

        # set the flag to True
        analyser.data_requirements[f"donor_z_{self.label}"].set_update_flag(True)
        analyser.data_requirements[f"acceptor_z_{self.label}"].set_update_flag(True)
        analyser.data_requirements[f"da_distance_{self.label}"].set_update_flag(True)
        analyser.data_requirements[f"hbond_angle_{self.label}"].set_update_flag(True)
        analyser.data_requirements[f"donor_cn_{self.label}"].set_update_flag(True)
        analyser.data_requirements[f"acceptor_cn_{self.label}"].set_update_flag(True)

    def _conclude(self, analyser: PlanarInterfaceAnalysisBase):
        box_length = analyser.universe.dimensions[analyser.axis]
        # within one cell length positive of z1.
        donor_z_unwrapped = getattr(analyser, f"donor_z_{self.label}")
        donor_z_wrapped = utils.mic_1d(
            donor_z_unwrapped - analyser.r_surf_ref[:, np.newaxis, np.newaxis],
            box_length=box_length,
            ref=box_length / 2,
        )
        acceptor_z_unwrapped = getattr(analyser, f"acceptor_z_{self.label}")
        acceptor_z_wrapped = utils.mic_1d(
            acceptor_z_unwrapped - analyser.r_surf_ref[:, np.newaxis, np.newaxis],
            box_length=box_length,
            ref=box_length / 2,
        )
        frame_index = np.tile(
            np.arange(analyser.n_frames).reshape(-1, 1, 1),
            (1, self.ag_oxygen.n_atoms, self.max_HB),
        )
        mask = ~np.isnan(donor_z_wrapped)
        # output is [frame_index, z_donor, z_acceptor, distance, angle, CN_donor, CN_acceptor]
        self.results.hbonds = np.stack(
            [
                frame_index[mask],
                donor_z_wrapped[mask],
                acceptor_z_wrapped[mask],
                getattr(analyser, f"da_distance_{self.label}")[mask],
                getattr(analyser, f"hbond_angle_{self.label}")[mask],
                getattr(analyser, f"donor_cn_{self.label}")[mask],
                getattr(analyser, f"acceptor_cn_{self.label}")[mask],
            ],
            axis=-1,
        )


class RadialCorrelationFunction(SingleAnalysis):
    def __init__(
        self,
        oxygen_sel: str = "name O",
        hydrogens_sel: str = "name H",
        label: str = "oxygen",
        d_bin: float = 0.1,
        cutoff: float = 4.5,
    ) -> None:
        super().__init__()
        self.oxygen_sel = oxygen_sel
        self.hydrogens_sel = hydrogens_sel
        self.label = label

        assert d_bin > 0, "Bin width must be greater than 0."
        self.d_bin = d_bin
        self.cutoff = cutoff

        self.ag_oxygen = None
        self.ag_hydrogen = None
        self.calculator = None
        self.corr_func = None

    def _prepare(self, analyser: PlanarInterfaceAnalysisBase):
        self.ag_oxygen = analyser.universe.select_atoms(self.oxygen_sel)
        self.ag_hydrogen = analyser.universe.select_atoms(self.hydrogens_sel)

        cell_vectors = geometry.cellpar_to_cell(analyser.universe.dimensions)
        cell_vectors = np.delete(cell_vectors, analyser.axis, axis=0)
        r_max = utils.calc_pbc_r_max(*cell_vectors)
        bins = int(r_max / self.d_bin)
        self.calculator = freud.density.CorrelationFunction(
            bins=bins, r_max=r_max / 1.01
        )
        self.corr_func = np.zeros((analyser.n_frames, 3, bins))

    def _single_frame(self, analyser: PlanarInterfaceAnalysisBase):
        ts_box = analyser._ts.dimensions
        box_length = ts_box[analyser.axis]

        # calculate mask based on self.cutoff
        ts_r_surf_lo = analyser.r_surf_lo[analyser._frame_index]
        ts_r_surf_hi = utils.mic_1d(
            analyser.r_surf_hi[analyser._frame_index] - ts_r_surf_lo,
            box_length,
            ref=box_length / 2,
        )
        ts_wrapped_r = utils.mic_1d(
            self.ag_oxygen.positions[:, analyser.axis] - ts_r_surf_lo,
            box_length,
            ref=box_length / 2,
        )
        # z distance between the oxygen atom and the surface
        ts_wrapped_r = np.min([ts_wrapped_r, ts_r_surf_hi - ts_wrapped_r], axis=0)
        mask = ts_wrapped_r < self.cutoff

        ts_selected_oxygen = self.ag_oxygen[mask]
        # print(ts_selected_oxygen.positions[:, analyser.axis])

        # Compute neighbor list of selected oxygen atoms
        nlist = (
            freud.AABBQuery.from_system(analyser._ts, 3)
            .query(ts_selected_oxygen.positions, {"r_max": 5.0})
            .toNeighborList()
        )

        ds = distance_array(
            ts_selected_oxygen.positions,
            self.ag_oxygen.positions,
            box=analyser._ts.dimensions,
        )
        min_vectors = minimize_vectors(
            self.ag_oxygen.positions[np.argsort(ds, axis=1)[:, 1]]
            - ts_selected_oxygen.positions,
            box=analyser._ts.dimensions,
        )
        # print(min_vectors)
        # calculate the angles
        angles = np.arccos(min_vectors[:, 0] / np.linalg.norm(min_vectors, axis=1))
        self.calculator.compute(
            system=analyser._ts,
            neighbors=nlist,
            values=analyser._ts.positions[:, analyser.axis],
            query_points=ts_selected_oxygen.positions,
            query_values=angles,
        )
        np.copyto(self.corr_func[analyser._frame_index, 0], self.calculator.correlation)

        ds = distance_array(
            ts_selected_oxygen.positions,
            self.ag_hydrogen.positions,
            box=analyser._ts.dimensions,
        )
        min_vectors = minimize_vectors(
            self.ag_hydrogen.positions[np.argsort(ds, axis=1)[:, 0]]
            - ts_selected_oxygen.positions,
            box=analyser._ts.dimensions,
        )
        angles = np.arccos(min_vectors[:, 0] / np.linalg.norm(min_vectors, axis=1))
        self.calculator.compute(
            system=analyser._ts,
            neighbors=nlist,
            values=analyser._ts.positions[:, analyser.axis],
            query_points=ts_selected_oxygen.positions,
            query_values=angles,
        )
        np.copyto(self.corr_func[analyser._frame_index, 1], self.calculator.correlation)

        min_vectors = minimize_vectors(
            self.ag_hydrogen.positions[np.argsort(ds, axis=1)[:, 1]]
            - ts_selected_oxygen.positions,
            box=analyser._ts.dimensions,
        )
        angles = np.arccos(min_vectors[:, 0] / np.linalg.norm(min_vectors, axis=1))
        self.calculator.compute(
            system=analyser._ts,
            neighbors=nlist,
            values=analyser._ts.positions[:, analyser.axis],
            query_points=ts_selected_oxygen.positions,
            query_values=angles,
        )
        np.copyto(self.corr_func[analyser._frame_index, 2], self.calculator.correlation)

    def _conclude(self, analyser: PlanarInterfaceAnalysisBase):
        self.results.bins = self.calculator.bin_centers
        corr_func = np.mean(self.corr_func, axis=0)
        # normalised corr_func to the first bin
        self.results.corr_func = corr_func / corr_func[:, 0][:, np.newaxis]
