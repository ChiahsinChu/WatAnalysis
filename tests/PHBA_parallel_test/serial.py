import numpy as np
import time
import MDAnalysis as mda
from MDAnalysis import transformations as trans

from WatAnalysis.hbonds.m_mda import PartialHBAnalysis

# load trajectory
u = mda.Universe("../input_data/interface.psf", "../input_data/trajectory.xyz")
dim = [16.869, 16.869, 41.478, 90, 90, 120]
transform = trans.boxdimensions.set_dimensions(dim)
u.trajectory.add_transformations(transform)

start = time.time()
hbonds = PartialHBAnalysis(universe=u,
                           hb_region=[0, 4.5],
                           surf_ids=[np.arange(871, 906),
                                     np.arange(691, 726)],
                           donors_sel=None,
                           hydrogens_sel="name H",
                           acceptors_sel="name O",
                           d_a_cutoff=3.0,
                           angle_cutoff_type='d_h_a',
                           angle_cutoff=150,
                           update_selections=False)
hbonds.run()
print(hbonds.results.hbonds.shape)
fmt = "\nWork Completed! Used Time: {:.3f} seconds"
print(fmt.format(time.time() - start))