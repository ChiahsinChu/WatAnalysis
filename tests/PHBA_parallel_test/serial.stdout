(808369, 9)

Work Completed! Used Time: 73.790 seconds
